tools/schemapi/schemapi.py

```diff
@@ -1279,10 +1528,93 @@ def _is_iterable(
     return not isinstance(obj, exclude) and isinstance(obj, Iterable)
 
 
+def _is_valid(spec: _JsonParameter, tp: type[SchemaBase], /) -> bool:
+    """
+    Return True if ``tp`` can be constructed from ``spec``.
+
+    Notes
+    -----
+    Don't use this if you need to know *details* of the errors in ``spec``..
+    """
+    return next(_validator(tp._schema, tp._rootschema).iter_errors(spec), None) is None
+
+
 def _passthrough(*args: Any, **kwds: Any) -> Any | dict[str, Any]:
     return args[0] if args else kwds
 
 
+def _freeze(val):
+    # NOTE: No longer referenced
+    # - Previously only called during tests
+    # - Not during any library code
+    if isinstance(val, dict):
+        return frozenset((k, _freeze(v)) for k, v in val.items())
+    elif isinstance(val, set):
+        return frozenset(_freeze(v) for v in val)
+    elif isinstance(val, (list, tuple)):
+        return tuple(_freeze(v) for v in val)
+    else:
+        return val
+
+
+def _hash_schema(
+    schema: _JsonParameter,
+    /,
+    *,
+    exclude: Iterable[str] = frozenset(
+        ("definitions", "title", "description", "$schema", "id")
+    ),
+) -> int:
+    """
+    Return the hash value for a ``schema``.
+
+    Parameters
+    ----------
+    schema
+        ``SchemaBase._schema``.
+    exclude
+        ``schema`` keys which are not considered when identifying equivalence.
+    """
+    if isinstance(schema, Mapping):
+        schema = {k: v for k, v in schema.items() if k not in exclude}
+    return hash(_HASH_ENCODER.encode(schema))
+
+
+def _subclasses(cls: type[TSchemaBase]) -> Iterator[type[TSchemaBase]]:
+    """
+    Breadth-first sequence of all classes which inherit from ``cls``.
+
+    Notes
+    -----
+    - `__subclasses__()` alone isn't helpful, as that is only immediate subclasses
+    - Deterministic
+    - Used for `SchemaBase` & `VegaLiteSchema`
+    - In practice, it provides an iterator over all classes in the schema below `VegaLiteSchema`
+        - The first one is `Root`
+    - The order itself, I don't think is important
+        - But probably important that it doesn't change
+        - Thinking they used an iterator so that the subclasses are evaluated after they have all been defined
+
+    - `Chart` seems to try to avoid calling this
+        - Using `TopLevelMixin.__subclasses__()` first if possible
+    - It is always called during `Chart.encode()`
+        - Chart.encode()
+        - altair.utils.core.infer_encoding_types
+        -  _ChannelCache.infer_encoding_types
+        - _ChannelCache._wrap_in_channel
+        - SchemaBase.from_dict (recursive, hot loop, validate =False, within a try/except)
+        - _FromDict(cls._default_wrapper_classes())
+        - schemapi._subclasses(schema.core.VegaLiteSchema)
```

Comment:
TODO: Tidy this up

Reply:
Resolved in [`fba3c46` (#3547)](https://github.com/vega/altair/pull/3547/commits/fba3c46c0aa69a99d731c6a39a90ce396eae8863)